signer/src/block_observer.rs

```diff
@@ -243,60 +247,104 @@ impl<C: Context, B> BlockObserver<C, B> {
         Ok(())
     }
 
-    /// Find the parent blocks from the given block that are also missing from our database
+    /// Set the sbtc start height, if it has not been set already.
+    async fn set_sbtc_start_height(&self) -> Result<(), Error> {
+        if self.context.state().is_sbtc_start_height_set() {
+            return Ok(());
+        }
+
+        let pox_info = self.context.get_stacks_client().get_pox_info().await?;
+        let nakamoto_start_height = pox_info
+            .nakamoto_start_height()
+            .ok_or(Error::MissingNakamotoStartHeight)?;
+
+        self.context
+            .state()
+            .set_sbtc_start_height(nakamoto_start_height);
+
+        Ok(())
+    }
+
+    /// Find the parent blocks from the given block that are also missing
+    /// from our database.
+    ///
+    /// # Notes
+    ///
+    /// This function does two things:
+    /// 1. Set the `sbtc_start_height` if it has not been set already. If
+    ///    it is not set, then we fetch the stacks nakamoto start height
+    ///    from stacks-core and use that value.
+    /// 2. Continually fetches block headers from bitcoin-core until it
+    ///    encounters a known block header or if the height of the block is
+    ///    less than or equal to the `sbtc_start_height`.
+    ///
+    /// If there are many unknown blocks then this function can take some
+    /// time. Since each header is 80 bytes, we should be able to fetch
+    /// headers for the entire bitcoin blockchain (~900k blocks at the time
+    /// of writing) into memory.
     #[tracing::instrument(skip_all, fields(%block_hash))]
-    async fn next_blocks_to_process(
+    pub async fn next_headers_to_process(
         &self,
-        mut block_hash: bitcoin::BlockHash,
-    ) -> Result<Vec<bitcoin::Block>, Error> {
-        let mut blocks = Vec::new();
+        mut block_hash: BlockHash,
+    ) -> Result<Vec<BitcoinBlockHeader>, Error> {
+        self.set_sbtc_start_height().await?;
 
-        for _ in 0..self.horizon {
-            if self.have_already_processed_block(&block_hash).await? {
-                tracing::debug!(%block_hash, "already processed block");
+        let start_height = self.context.state().get_sbtc_start_height();
+        let mut headers = std::collections::VecDeque::new();
+        let db = self.context.get_storage();
+        let bitcoin_client = self.context.get_bitcoin_client();
+
+        while !db.is_known_bitcoin_block_hash(&block_hash.into()).await? {
+            let Some(header) = bitcoin_client.get_block_header(&block_hash).await? else {
+                tracing::warn!(%block_hash, "bitcoin-core does not know about block header");
+                return Ok(headers.into());
```

Comment:
Is this an `ok` situation? Could it be the case (eg, bitcoin node restarted/catching up while signer stays up) that we end up with gaps in the chain?
If it's `ok`, we can just `break`.

Reply:
No, it is not! I need to return an empty `Vec`.

Reply:
I see this is resolved in [5d36eea](https://github.com/stacks-network/sbtc/pull/1093/commits/5d36eea859aab5b806f9e4706d6fa99c9d87d19a):

```
let Some(header) = bitcoin_client.get_block_header(&block_hash).await? else {
    tracing::error!(%block_hash, "bitcoin-core does not know about block header");
    return Err(Error::BitcoinCoreUnknownBlockHeader(block_hash));
};
```

Reply:
Fixed in https://github.com/stacks-network/sbtc/pull/1093/commits/5d36eea859aab5b806f9e4706d6fa99c9d87d19a.